### DISCUSSION TITLE
Let's discuss the formatting behavior

### DISCUSSION THREAD

**Discussion by introt** [zim-desktop-wiki/zim-desktop-wiki#1731] [General]
Zim's text formatting is quite quirky. Let's discuss it here, one example per comment. Please include a description of the behavior of other editors, when applicable.

> Examples are to be formatted as quotes, with the cursor represented by a pipe character: |

**Comment by introt** [comment]
### Exhibit A: beginning of line

> |hello there

Pressing Ctrl+B bolds 'hello'; pressing it another time doesn't unbold it. Other formats like italics and emphasis can similarly be applied on top, but not removed. However, heading styles can be toggled on and off.

GitHub's online editor allows toggling the formatting of the first word on and off, while LibreOffice toggles the active style without affecting the existing text.

**Reply by introt** [reply]
I'd prefer the approach chosen by LO, as neither Zim nor GH allow one to easily add a  

> **Bold addition in front of** |existing text;

GH at least places the cursor inside the bold formatting, but with Zim you end up bolding the existing first word while your addition stays unformatted.

**Reply by tehkonst** [reply]
> Pressing Ctrl+B bolds 'hello'; pressing it another time doesn't unbold it.

Can't confirm. I've been using zim for several years and it has a lot of bugs. But boldness works as expected (on my linux dist at least)

**Reply by tscahill** [reply]
As tehkonst reports. I'm running zim 0.74.2 on manjaro linux. Ctrl-B toggles bold off and on.

**Reply by introt** [reply]
Weird - here the first Ctrl+B bolds "hello" without turning on bold (typed text isn't bold) and the subsequent presses toggle the active formatting on and off, but the "hello" stays bold. Ctrl+Z unbolds the "hello" and moves the cursor to the end of the word (which it shouldn't). What's your distro @tehkonst and what happens to the word after the cursor on your box @tscahill?

**Reply by introt** [reply]
Here's how the test setup looks like in the beginning:
![image](https://user-images.githubusercontent.com/26069787/143679634-ea070db0-0df7-4148-bf7d-faa8a503d459.png)

Here's the view after pressing Ctrl+B once; the "hello" goes bold, but bolding doesn't get toggled on:
![image](https://user-images.githubusercontent.com/26069787/143679671-5374025a-176b-41d9-8fd4-7a9dcebeb24d.png)

After a second press the bold formatting is toggled on, but the "hello" stays bold (my tiny window also grew for some reason):
![image](https://user-images.githubusercontent.com/26069787/143679659-87074a8b-c8c3-4ead-a5c6-ff6c23cd9941.png)

Subsequent presses just toggle the active formatting without affecting the "hello", eg. the view after the third press is identical to the one after the first one, as is the one after the fifth and so on.

Here's what it looks like after pressing Ctrl+Z:
![image](https://user-images.githubusercontent.com/26069787/143679768-995eb2d3-97fb-48dc-ad7c-b3072144545e.png)

**Reply by Self-Perfection** [reply]
I can reproduce what introt showed. I use Zim 0.74.1 on Arch linux with KDE.

Note that it is important to place cursor before the first char of a word.

@introt there lots of small nagging edge cases in Zim text editor. You just can't simply reverse an action by executing its reversed pair. For instance I have rather frequently issues with line sorter plugin which might merge moved line with header instead of swapping them. I just learned to hit undo if I got something undesirable in Zim.

**Reply by introt** [reply]
> You just can't simply reverse an action by executing its reversed pair

Not sure if you mean in Zim or in general; the weird thing here is that you totally can toggle the boldness of a word by having the cursor inside it while hitting Ctrl+B.

> I just learned to hit undo if I got something undesirable in Zim.

The problem is that undo doesn't reliably revert to the previous state, as seen above (the cursor ends up after the word).

I'm currently too busy to dig into the code, but it'd be helpful to have these edge cases neatly cataloged and ready for review by whoever happens to get around to it first - so please do leave a comment about any you happen to come across, with steps to reproduce if possible. Thanks!

**Reply by tscahill** [reply]
**My mistake.** It never occured to me to attempt formating of text without first selecting the entire string of text I wanted to format. My result was of a different process. 

In the procedure introt describes, my experience matches his. With the cursor just ahead of the text, it toggles bold on but not off. (I just realized the world has been spelling proceedure wrong).

**Reply by tehkonst** [reply]
> the weird thing here is that you totally can toggle the boldness of a word by having the cursor inside it while hitting Ctrl+B

It's convenient. On github here it works the same way and in Libreoffice Writer. So I guess it's not weird it's the most expected behavior.

**Reply by tehkonst** [reply]
@introt
Sorry, I didn't understand you at first. Never thought of putting the cursor at the beginning for the formatting.

**Comment by introt** [comment]
### Exhibit B: selected formats don't stay active when erasing

> I wish to bold the next word, but I mistyped the first letter: |

After pressing Ctrl+B, the next letter you type will be bolded. However, if you erase it, you end up with nonformatted text; toggling a formatting doesn't place an invisible marker to mark where it starts, so erasing the first bolded letter erases all traces of bolding.

In LibreOffice, the selected styles would remain active unless the preceding character (space in this example) is erased or the cursor is moved (there isn't a literal "invisible marker" either. GH's MD editor places literal `**`s which don't go away).

**Reply by introt** [reply]
This is the number one editing flow breaker I have right now and I find it mildly infuriating.

**Comment by vossenjp** [comment]
On 11/27/21 10:51 AM, tscahill wrote:
> *My mistake.* It never occured to me to attempt formating of text without first selecting the entire string of text I wanted to format. In the setup introt describes, my experience matches his. With the cursor just ahead of the text, it toggles bold on but not off.

You might want to look at "Zim > Preferences > Editing: Automatically select the current word when you apply formatting" and see if that makes things better.

**Reply by introt** [reply]
Thanks for naming the feature! According to the docs[0] it should behave the same regardless of cursor position so I'm filing an issue.

0. https://github.com/zim-desktop-wiki/zim-desktop-wiki/blob/4c68661c758457a4cbb08a99806c3958f2b5ea96/zim/gui/pageview/__init__.py#L7285

**Comment by introt** [comment]
# Exhibit C: indenting bullets

To turn one or more lines into a bullet list, one can select any part of a line or multiple lines and press `*`; however, to (un)indent these bullets with tab/backspace, your selection must begin from next to the bullet or the selected text gets erased.

![image](https://user-images.githubusercontent.com/26069787/145873358-381632ae-3f35-4257-8942-11bc98d69718.png)
_An innocuous looking selection_

![image](https://user-images.githubusercontent.com/26069787/145873382-add37422-79f3-4d55-bb1c-d14835e98045.png)
_"Oops"_

![image](https://user-images.githubusercontent.com/26069787/145873460-3d23077d-aa85-4cd6-8d05-4b2bf73f7e47.png)
_The "correct" way to select lines_

**Reply by introt** [reply]
I'd like this to work like `*`, but that'd make deleting selected text with backspace impossible. To work around this issue I propose switching the unindent shortcut to `Shift` + `Tab` , as used by some other applications.

**Comment by TwiggyWan** [comment]
Hello may I submit an exhibit too? I'm considering opening a feature request issue for it.

# Line sorter does not respect list indent
If I use Line sorter on this example : 

```
[ ] DEF 
	[ ] Z
	[ ] W
[ ] ABC
	[ ] GHI
	[ ] JK

```

I get : 

```
[ ] ABC
[ ] DEF 
	[ ] GHI
	[ ] JK
	[ ] W
	[ ] Z
	
```

But I expected : 

```

[ ] ABC
	[ ] GHI
	[ ] JK

[ ] DEF 
	[ ] W
	[ ] Z
```

Same behavior if from the orig snippet I select this and press Ctrl + Down
![image](https://github.com/user-attachments/assets/44f334ec-13e6-4bc8-a49c-efd9d2e5dccb)

![image](https://github.com/user-attachments/assets/7130567c-d17f-4668-9c21-18572871ddac)

I expected : 
![image](https://github.com/user-attachments/assets/43c12c31-6597-4c51-80f6-db9f51266760)
